network/transport/v2/handlers.go

```diff
@@ -151,10 +158,114 @@ func (p protocol) handleGossip(peer transport.Peer, msg *Gossip) error {
 	if len(refs) > 0 {
 		// TODO swap for trace logging
 		log.Logger().Infof("received %d new transaction references via Gossip", len(refs))
+		p.sendTransactionListQuery(peer.ID, refs)
 	}
 	p.gManager.GossipReceived(peer.ID, refs...)
 	// TODO compare xor
 	// TODO send new message
 
 	return nil
 }
+
+func (p *protocol) handleTransactionList(peer transport.Peer, envelope *Envelope_TransactionList) error {
+	msg := envelope.TransactionList
+	cid, err := parseConversationID(msg.ConversationID)
+	if err != nil {
+		return fmt.Errorf("invalid conversationID: %w", err)
+	}
+	// TODO convert to trace logging
+	log.Logger().Infof("handling handleTransactionList from peer (peer=%s, conversationID=%s)", peer.ID.String(), cid.String())
+
+	// check if response matches earlier request
+	if err = p.cMan.check(envelope); err != nil {
+		return err
+	}
+
+	ctx := context.Background()
+	for _, tx := range msg.Transactions {
+		transactionRef := hash.FromSlice(tx.Hash)
+		transaction, err := dag.ParseTransaction(tx.Data)
+		if err != nil {
+			return fmt.Errorf("received transaction is invalid (peer=%s, ref=%s): %w", peer, transactionRef, err)
+		}
+
+		present, err := p.state.IsPresent(ctx, transactionRef)
+		if !present {
+			// TODO does this always trigger fetching missing payloads? (through observer on DAG) Prolly not for v2
+			if err = p.state.Add(ctx, transaction, tx.Payload); err != nil {
+				return fmt.Errorf("unable to add received transaction to DAG (tx=%s): %w", transaction.Ref(), err)
+			}
+		}
+	}
+
+	// TODO done is not called on the cMan since we don't know if we received all chunks
+
+	return nil
+}
+
+func (p *protocol) handleTransactionListQuery(peer transport.Peer, msg *TransactionListQuery) error {
+	requestedRefs := make([]hash.SHA256Hash, len(msg.Refs))
+	transactions := make([]*Transaction, 0)
+	unsorted := make([]dag.Transaction, 0)
+
+	cid, err := parseConversationID(msg.ConversationID)
+	if err != nil {
+		return err
+	}
+	// TODO convert to trace logging
+	log.Logger().Infof("handling transactionListQuery from peer (peer=%s, conversationID=%s)", peer.ID.String(), cid.String())
+
+	for i, refBytes := range msg.Refs {
+		requestedRefs[i] = hash.FromSlice(refBytes)
+	}
+
+	if len(requestedRefs) == 0 {
+		log.Logger().Warnf("peer sent request for 0 transactions (peer=%s)", peer.ID)
+		return nil
+	}
+
+	ctx := context.Background()
+
+	// first retrieve all transactions, this is needed to sort them on LC value
+	for _, ref := range requestedRefs {
+		transaction, err := p.state.GetTransaction(ctx, ref)
+		if err != nil {
+			return err
+		}
+		// If a transaction is not present, we stop any further transaction gathering.
```

Comment:
How is `requestedRefs` ordered? Depending on order (LC ascending/descending, random order, some other order) it makes sense to stop gathering TXs or not. If the order is random, there could be a very recent TX halfway that we  just don't have yet (while the rest could still be sent back).

Should it be spec'd?

Reply:
order is unknown, but this query is sent as a direct result of the state this node has. So logically it's almost impossible to miss anything. To make it a bit more resilient, the node can just send whatever it has.